git_repo/services/ext/gitbucket.py

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
+
+import logging
+log = logging.getLogger('git_repo.gitbucket')
+
+from ..service import register_target
+from .github import GithubService
+
+import github3
+
+@register_target('bucket', 'gitbucket')
+class GitbucketService(GithubService):
+    fqdn = "localhost"
+    port = 8080
+    def __init__(self, *args, **kwarg):
+        super(GitbucketService, self).__init__(*args, **kwarg)
+
+    def connect(self):
+        self.gh = github3.GitHubEnterprise(self.build_url(self))
+        super(GitbucketService, self).connect()
+
+    @classmethod
+    def get_auth_token(cls, login, password, prompt=None):
+        print("Please edit _YOUR_ACCESS_TOKEN_ to actual token in ~/.gitconfig or ~/.git/config after.")
+        print("And add ssh-url = ssh://git@{}:_YOUR_SSH_PORT_".format(cls.fqdn))
+        return "_YOUR_ACCESS_TOKEN_"
+        ## this code maybe works when GitBucket supports add access token API.
+        #print("build_url: ", cls.build_url(cls))
+        #import platform
+        #gh = github3.GitHubEnterprise(cls.build_url(cls))
+        #gh.login(login, password, two_factor_callback=lambda: prompt('2FA code> '))
+        #try:
+        #    auth = gh.authorize(login, password,
+        #            scopes=[ 'repo', 'delete_repo', 'gist' ],
+        #            note='git-repo2 token used on {}'.format(platform.node()),
+        #            note_url='https://github.com/guyzmo/git-repo')
+        #    return auth.token
+        #except github3.models.GitHubError as err:
+        #    if len(err.args) > 0 and 422 == err.args[0].status_code:
+        #        raise ResourceExistsError("A token already exist for this machine on your github account.")
+        #    else:
+        #        raise err
+
+    def format_path(self, repository, namespace=None, rw=False):
+        path = super(GitbucketService, self).format_path(repository, namespace, rw)
+        if not path.endswith(".git"):
+            path = "{}.git".format(path)
```

Comment:
I know it.

In GitBucket, 
for ssh access, `.git` is required.
for http(s) access, GitBucket's formal http(s) URL format is `http://server:port/git/owner/repo.git`. 
But it also can access with:

- `http://server:port/git/owner/repo`
- `http://server:port/owner/repo.git`

But it can't access by `http://server:port/owner/repo`. Then it always add `.git`.

Recently, I found problem with git-2.12. I wrote it in https://github.com/gitbucket/gitbucket/issues/1552.

In GitBucket-4.12, `http://server:port/owner/repo.git` returns redirect to `http://server:port/git/owner/repo.git`. git-2.12 doesn't allow redirect url.
 so I will rewrite `format_path` such as:
- never call `super(...).format_path`
- returns `ssh://user@server:port/owner/repo.git` for rw=True
- returns `http://server:port/git/owner/repo.git` for rw=False